File: DesignPatterns/5-templateMethodPattern/golang/littleApprentice.go
/*
 * File: littleApprentice.go
 * Created Date: 2023-03-13 03:00:07
 * Author: ysj
 * Description:  模板方式模式——小徒弟铸剑
 */

package main

import "fmt"

type LittleApprentice struct {
	ForgeSwordBase
}

// 初始化
func NewLittleApprentice() ForgeSwordTemplate {
	return &LittleApprentice{}
}

func (la *LittleApprentice) ForgeSword() {
	la.makeClayModel()
	la.dispenseMaterials()
	la.smeltingMaterials()
	la.waterShaping()
	la.repairProcessing()

	la.result()
}

func (la *LittleApprentice) dispenseMaterials() {
	fmt.Println("调剂材料比例不均 +10分")
	la.Score += 10
}

func (la *LittleApprentice) smeltingMaterials() {
	fmt.Println("熔炼原料火候不够 +10分")
	la.Score += 10
}

File: DesignPatterns/5-templateMethodPattern/python/little_apprentice.py
#!/usr/bin/env python3
# -*- coding:utf-8 -*-
###
# File: little_apprentice.py
# Created Date: 2023-03-13 03:23:08
# Author: ysj
# Description:  模板方式模式——小徒弟铸剑
###

from template import ForgeSwordTemplate


class LittleApprenticeForgeSword(ForgeSwordTemplate):
    def make_clay_model(self):
        '''制做泥范, 每道工序20分'''
        super().make_clay_model()

    def dispense_materials(self):
        print("调剂材料比例不均 +10分")
        self.score += 10

    def smelting_materials(self):
        print("熔炼原料火候不够 +10分")
        self.score += 10

    def water_shaping(self):
        '''浇灌成形, 每道工序20分'''
        super().water_shaping()

    def repair_processing(self):
        '''修治加工, 每道工序20分'''
        super().repair_processing()
